service/src/tests/mocks/parentchain_api_mock.rs

```diff
@@ -0,0 +1,72 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use itp_api_client_extensions::{ApiResult, ChainApi, StorageProof};
+use itp_test::builders::{
+	parentchain_block_builder::ParentchainBlockBuilder,
+	parentchain_header_builder::ParentchainHeaderBuilder,
+};
+use itp_types::{SignedBlock, H256};
+use sp_finality_grandpa::AuthorityList;
+use std::cmp::min;
+
+pub struct ParentchainApiMock {
+	parentchain: Vec<SignedBlock>,
+}
+
+impl ParentchainApiMock {
+	pub(crate) fn new() -> Self {
+		ParentchainApiMock { parentchain: Vec::new() }
+	}
+
+	// can this be combined with the constructor? Does it make sense?
+	pub fn create_mock_parentchain(&mut self, len: u32) {
+		for n in 1..=len {
+			let header = ParentchainHeaderBuilder::default().with_number(n).build();
+			let block = ParentchainBlockBuilder::default().with_header(header).build_signed();
+			self.parentchain.push(block);
+		}
+	}
+}
+
+impl ChainApi for ParentchainApiMock {
+	fn last_finalized_block(&self) -> ApiResult<Option<SignedBlock>> {
+		let i = self.parentchain.len();
+		ApiResult::Ok(Some(self.parentchain[i - 1].clone()))
+	}
+
+	fn signed_block(&self, _hash: Option<H256>) -> ApiResult<Option<SignedBlock>> {
+		todo!()
+	}
+
+	fn get_blocks(&self, from: u32, to: u32) -> ApiResult<Vec<SignedBlock>> {
+		let mut blocks = Vec::new();
+		let len = self.parentchain.len();
+		for n in from as usize..=min(to as usize, len - 1) {
+			blocks.push(self.parentchain[n].clone())
+		}
```

Comment:
This is also very 'old'-style of code, using indices and pushing vectors. In Rust we prefer to approach this from a more functional style, as is done in Java with streams, or LINQ in C#.
Something along the lines of (I haven't checked if it compiles, and maybe the `skip` has to be `from - 1` (but also with `checked_sub`)):
```
let num_elements = to.checked_sub(from)?;
let blocks = self.parentchain.iter().skip(from as usize).take(num_elements as usize).cloned().collect();
```